postprocessing/Postprocessing.java

```diff
@@ -0,0 +1,258 @@
+import us.shandian.giga.io.CircularFileWriter;
+import us.shandian.giga.io.CircularFileWriter.OffsetChecker;
+import us.shandian.giga.io.ProgressReport;
+
+import static us.shandian.giga.get.DownloadMission.ERROR_NOTHING;
+import static us.shandian.giga.get.DownloadMission.ERROR_POSTPROCESSING;
+import static us.shandian.giga.get.DownloadMission.ERROR_POSTPROCESSING_HOLD;
+
+public abstract class Postprocessing implements Serializable {
+
+    static transient final byte OK_RESULT = ERROR_NOTHING;
+
+    public transient static final String ALGORITHM_TTML_CONVERTER = "ttml";
+    public transient static final String ALGORITHM_WEBM_MUXER = "webm";
+    public transient static final String ALGORITHM_MP4_FROM_DASH_MUXER = "mp4D-mp4";
+    public transient static final String ALGORITHM_M4A_NO_DASH = "mp4D-m4a";
+    public transient static final String ALGORITHM_OGG_FROM_WEBM_DEMUXER = "webm-ogg-d";
+
+    public static Postprocessing getAlgorithm(@NonNull String algorithmName, String[] args) {
+        Postprocessing instance;
+
+        switch (algorithmName) {
+            case ALGORITHM_TTML_CONVERTER:
+                instance = new TtmlConverter();
+                break;
+            case ALGORITHM_WEBM_MUXER:
+                instance = new WebMMuxer();
+                break;
+            case ALGORITHM_MP4_FROM_DASH_MUXER:
+                instance = new Mp4FromDashMuxer();
+                break;
+            case ALGORITHM_M4A_NO_DASH:
+                instance = new M4aNoDash();
+                break;
+            case ALGORITHM_OGG_FROM_WEBM_DEMUXER:
+                instance = new OggFromWebmDemuxer();
+                break;
+            /*case "example-algorithm":
+                instance = new ExampleAlgorithm();*/
+            default:
+                throw new UnsupportedOperationException("Unimplemented post-processing algorithm: " + algorithmName);
+        }
+
+        instance.args = args;
+        return instance;
+    }
+
+    /**
+     * Get a boolean value that indicate if the given algorithm work on the same
+     * file
+     */
+    public boolean worksOnSameFile;
+
+    /**
+     * Indicates whether the selected algorithm needs space reserved at the beginning of the file
+     */
+    public boolean reserveSpace;
+
+    /**
+     * Gets the given algorithm short name
+     */
+    private final String name;
+
+
+    private String[] args;
+
+    private transient DownloadMission mission;
+
+    private transient File tempFile;
+
+    Postprocessing(boolean reserveSpace, boolean worksOnSameFile, String algorithmName) {
+        this.reserveSpace = reserveSpace;
+        this.worksOnSameFile = worksOnSameFile;
+        this.name = algorithmName;// for debugging only
+    }
+
+    public void setTemporalDir(@NonNull File directory) {
+        long rnd = (int) (Math.random() * 100000.0f);
+        tempFile = new File(directory, rnd + "_" + System.nanoTime() + ".tmp");
+    }
```

Comment:
<div>


<div id="suggestion">
<div id="issue"><b>Consider more secure temp file creation</b></div>
<div id="fix">

The temporary file name generation in `setTemporalDir()` may be vulnerable to race conditions. Consider using `Files.createTempFile()` instead of concatenating random numbers.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -91,4 +91,5 @@
  public void setTemporalDir(@NonNull File directory) {
 -    long rnd = (int) (Math.random() * 100000.0f);
 -    tempFile = new File(directory, rnd + "_" + System.nanoTime() + ".tmp");
 +    try {
 +        tempFile = Files.createTempFile(directory.toPath(), "download", ".tmp").toFile();
 +    } catch (IOException e) {
 +        Log.e("Postprocessing", "Failed to create temp file", e);
 +    }
  }
```
</div>
</details>
</div>


<small><i>Code Review Run <a href=https://github.com/OBrutus/youtube-downloader/pull/9#issuecomment-2464377435>#3b9ab6</a></i></small>
</div>

---
Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged